### DISCUSSION TITLE
Comparison between pattern and meander?

### DISCUSSION THREAD

**Discussion by dgr** [pangloss/pattern#7] [General]
I was watching the London Clojurians talk on pattern and couldn't help thinking about [meander](https://github.com/noprompt/meander). There are a lot of similarities between the two libraries. It would be interesting to compare the two and understand the similarities and differences. What does pattern do that meander doesn't, and vice-versa?

**Comment by pangloss** [comment]
I'm only familiar in passing with Meander and when I look at it I don't see a lot of similarity. Perhaps it would be better if you could ask more specific questions about features or capabilities.

I've been asked this a few times though, so digging up past responses, here are some general answers:

I don’t think there is much similarity with Meander, other than some very simple cases being expressed similarly. Meander seems to be macro based; pattern is purely functional at its core, with a few macros for ease of use. Meander seems to intermix pattern matching and transformation; pattern strictly separates them. 

Probably you don't want to learn pattern by thinking about it in terms of how Meander does things since the approach is very different.

Pattern is highly extensible and modular. I have many simple transformations I've written in Pattern that are just a simple match and replace, or even just a simple match, but I also have systems of hundreds of patterns that work together to do a single highly complex transformation (think large compiler pass). In the compiler pass I've also defined 40 (!) custom domain-specific matchers that I can use within any/all of those patterns.

**Comment by dgr** [comment]
Hm. OK. Superficially, I see a lot of overlap (pattern matching, replacement, rewriting, etc.). Functional vs. macros is just an implementation difference and not really relevant, other than possibly some performance differences. I don't know enough about Pattern at this point to know whether that overlap I see is just superficial or more meaningful, so that's what I was looking for.

**Reply by pangloss** [reply]
I think functional vs macros is actually a substantial difference once you move beyond the simpler use cases. Being able to pass compiled rules and pattern matchers around as functions enables a wider range of use cases.
